### PR TITLE
add support for string option bounds

### PR DESCRIPTION
**Pull request by caneleex**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [ ] Internal code
- [x] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

this PR adds support for setting the minimum and maximum length for `STRING` options: https://github.com/discord/discord-api-docs/pull/5143

it's been a while since I touched JDA/Java so there may be things I forgot to add/change.

### CODE REVIEW

**Comment by MinnDevelopment** [comment]
Could we allow to use `setRequiredRange` for these as well?

https://github.com/DV8FromTheWorld/JDA/blob/master/src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java#L559

**Comment by caneleex** [comment]
> Could we allow to use `setRequiredRange` for these as well?
> 
> [`master`/src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java#L559](https://github.com/DV8FromTheWorld/JDA/blob/master/src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java?rgh-link-date=2022-07-02T10%3A06%3A36Z#L559)

something like `OptionData#setRequiredLengthRange`?

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (caneleex) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:641]: would it make sense to add a bound check for the maximum value here as well?

**Comment by MinnDevelopment** [comment]
> something like OptionData#setRequiredLengthRange?

no, just the same method with a conditional check for type.

**Comment by caneleex** [comment]
> > something like OptionData#setRequiredLengthRange?
> 
> no, just the same method with a conditional check for type.

I feel like that could be a bit confusing. introducing a new method would get rid of having to explain 2 behaviors of that method in the docs + long -> int conversions. making an overload which'd accept ints instead would probably confuse/annoy users as well - if they wanted to set the range for an integer option, they'd have to use **L**ongs specifically otherwise it'd clash with the ints overload and throw, or am I misunderstanding here?

**Comment by MinnDevelopment** [comment]
> I feel like that could be a bit confusing

I think it's more confusing for it not to work, which is why I'm suggesting it.

> making an overload which'd accept ints instead 

I don't think we need one?

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (Xirado) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:696]:
```suggestion
     *             <li>{@link OptionType type} of this option is {@link OptionType#STRING STRING}</li>
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (Xirado) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:715]: This should probably also check if `minLength` is greater than `maxLength`

**Review by Xirado — CHANGES_REQUESTED** [review]

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (caneleex) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:696]: this is correct.

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (caneleex) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:715]: this isn't done in the range methods either, so I'd much rather leave this as is and maybe make a followup PR if you really wanted to and change it for both range and length methods.

Comment (DV8FromTheWorld) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:715]: Currently this is done during the serialization for min and max value.

Comment (DV8FromTheWorld) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:715]: Well, that's interesting. I definitely remember reading code that was doing the min/max comparison at serialization time because I remember deleting a comment i left on the min/max PR about adding that check in on the setX methods themselves but realized we were doing it just before serialization.

We definitely need to add these in. Them not being there for the others doesn't mean we shouldn't add it for these. I see no reason why we wouldn't add them tbh.

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (caneleex) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:655]: I'll add the check. however, according to docs, quote:

> The value of `min_length` must be greater or equal to `0`

this check is correct.

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (caneleex) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:77]: the name you proposed doesn't reflect what the field is for. the minimum length for the *maximum* length (`max_length`) should be greater than or equal to this value.

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (caneleex) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:715]: where do I put it? putting the check into the `setXLength` methods doesn't really make sense because you can't know the order of these calls, and I doubt it's good practice to throw in `toData()`.

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:80]: Do we really need a constant for this?

Comment (caneleex) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:80]: it's your call really.

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (Xirado) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:843]: That's how i did it in `TextInput`:
https://github.com/DV8FromTheWorld/JDA/blob/6e8c030fbe000c16224f658b1e37dffabe5a9383/src/main/java/net/dv8tion/jda/api/interactions/components/text/TextInput.java#L341-L350

In the cases where users use `setMaxLength()` and `setMinLength()`, it's checked in the `build()` method.

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:783]:
```suggestion
        Checks.notNegative(minLength, "Min length");
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:809]:
```suggestion
        Checks.positive(maxLength, "Max length");
        Checks.check(maxLength <= MAX_STRING_OPTION_LENGTH, "Max length must not be greater than %d. Provided: %d", MAX_STRING_OPTION_LENGTH, maxLength);
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:837]:
```suggestion
        Checks.notNegative(minLength, "Min length");
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:839]:
```suggestion
        Checks.check(maxLength <= MAX_STRING_OPTION_LENGTH,
                "Max length must not be greater than %d. Provided: %d", MAX_STRING_OPTION_LENGTH, maxLength);
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:840]:
```suggestion
        Checks.check(minLength <= maxLength, "Min length must not be greater than max length. Provided: %d > %d", minLength, maxLength);
        this.minLength = minLength;
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:80]: I think we can remove this now

**Review by MinnDevelopment — APPROVED** [review]

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:769]: Oh I thought we killed this way of defining `IllegalArgumentException`.
It really should be more like
```
@throws IllegalArgumentException
     *         <ul>
     *             <li>If the {@link OptionType type of this option} is not {@link OptionType#STRING STRING}</li>
     *             <li>If {@code minLength} is less than to {@code 0}</li>
     *         </ul>
``` 

I guess we don't _have_ to do it in this PR, but this is the pattern we've moved to in JDA in every other place.

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:836]: Is there a reason we re-implement these checks instead of doing:
```suggestion
        Checks.check(minLength <= maxLength, "Min length must not be greater than max length. Provided: %d > %d", minLength, maxLength);
        this.setMinLength(minLength);
        this.setMaxLength(maxLength);
```

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java
Comment (DV8FromTheWorld) [reply src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:769]: Ah, tracked this down. We changed it in a PR that never got merged.
I updated and merged that PR, so this language should be changed to match the new system before we merge.

**Review by DV8FromTheWorld — APPROVED** [review]

**Pull request merged**
